src/rnnlm/rnnlm-training.cc

```diff
@@ -0,0 +1,587 @@
+  }
+
+  if (supply_deriv && nnet != NULL) {
+    // the derivative on the real output
+    vector<vector<BaseFloat> > output_deriv(k);
+
+    for (int i = 0; i < k; i++) {
+      output_deriv[i].push_back(1);
+    }
+
+    // the derivative after the affine layer (before the nonlin)
+
+    // the derivative of the 'nnet3' part
+    CuMatrix<BaseFloat> input_deriv(output_0_gpu.NumRows(),
+                                    output_0_gpu.NumCols(),
+                                    kSetZero);
+
+    CuMatrix<BaseFloat> place_holder;
+    output_projection->Backprop(indexes, output_0_gpu, place_holder,
+                                output_deriv, nnet->output_projection_,
+                                &input_deriv);
+
+//    CuMatrix<BaseFloat> cu_input_deriv(input_deriv);
+
+    computer->AcceptOutputDeriv(output_name, &input_deriv);
+  }
+}
+
+void LmNnetSamplingTrainer::ComputeObjectiveFunctionSample(
+                              const vector<BaseFloat> &unigram,
+                              const GeneralMatrix &supervision,
+                              ObjectiveType objective_type,
+                              const std::string &output_name,
+                              bool supply_deriv,
+                              NnetComputer *computer,
+                              BaseFloat *tot_weight,
+                              BaseFloat *tot_objf,
+                              const LmOutputComponent *output_projection,
+                              Matrix<BaseFloat> *new_output,
+                              LmNnet *nnet) {
+  const CuMatrixBase<BaseFloat> &output_0_gpu = computer->GetOutput(output_name);
+//  Matrix<BaseFloat> output_0(output_0_gpu.NumRows(), output_0_gpu.NumCols());
+//  output_0_gpu.CopyToMat(&output_0);
+  new_output->Resize(output_0_gpu.NumRows(), output_0_gpu.NumCols());
+  output_0_gpu.CopyToMat(new_output);
+  int k = supervision.NumRows();
+
+  KALDI_ASSERT(supervision.Type() == kSparseMatrix);
+  const SparseMatrix<BaseFloat> &post = supervision.GetSparseMatrix();
+
+  vector<int> samples = Select(unigram, k);
+  vector<vector<int> > indexes(k);
+  for (int i = 0; i < k; i++) {
+    indexes[i] = samples;
+
+    const SparseVector<BaseFloat> &sv = post.Row(i);                              
+    int non_zero_index = -1;                                                    
+    sv.Max(&non_zero_index); 
+    indexes[i].push_back(non_zero_index);
+  }
+
+  vector<vector<BaseFloat> > out;
+
+  output_projection->Propagate(*new_output, indexes, &out);
+
+  *tot_weight = post.Sum();
+  *tot_objf = 0;
+  for (int i = 0; i < k; i++) {
+    KALDI_ASSERT(out[i].size() == k + 1);
+//    KALDI_LOG << "out-" << i << " is " << out[i][k];
+    *tot_objf += out[i][k]; // last one (k) is the correct lable
+    for (int j = 0; j < k; j++) {
+//      KALDI_ASSERT(out[i][j] == out[i][j]);
+//      KALDI_LOG << "out-" << i << " " << j << " is " << out[i][j];
+      *tot_objf -= exp(out[i][j]) / unigram[indexes[i][j]];
```

Comment:
Hmm right, thanks for pointing out this mistake. This might be why I'm getting very bad training likelihood so far...

I will fix all the issues that you mentioned in the review. 

One of the logs for an actual training run is here 
/export/b02/hxu/kaldi/egs/ami/s5/debug-rnnlm-rnn-0.001-0.0001-1.03-128-200-8--sample/log/train.rnnlm.4.log